### PR TITLE
Timeslot Selection

### PR DESCRIPTION
**Pull request by siongesteban**
@shannon-coding Here's the initial PR for the timeslot selection which is ready for integration. I already created a store for the event and selected timeslot with `zustand`. Please check `useEventStore.ts` and `useTimeslotStore.ts` under `src/fullPageBookingForm/Hooks`.

### CODE REVIEW

File: src/fullPageBookingForm/Components/Common/Calendar/Calendar.tsx [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+  year,
+  date,
+  loading,
+  startOfWeek = "Su",
+  dateIsDisabled,
+  onDateChange,
+  onMonthChange,
+  onYearChange,
+}) => {
+  const [currentMonth, setCurrentMonth] = useState(
```
Comment (shannon-coding) [inline src/fullPageBookingForm/Components/Common/Calendar/Calendar.tsx:37]: Did you not want to lift this state of the component? It looks like we're tracking state locally as well as at the parent level.

File: src/fullPageBookingForm/Components/TimeslotCard/TimeslotCard.scss [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  gap: 30px;
+
+  .timeslot-card__details__time {
+    margin-bottom: 6px;
+  }
+
+  .timeslot-card__button {
+    width: 126px;
+  }
+  // .timeslot-card__details__pricing {
```
Comment (shannon-coding) [inline src/fullPageBookingForm/Components/TimeslotCard/TimeslotCard.scss:24]: Remove this.

File: src/fullPageBookingForm/Components/Views/OrderDetails/OrderDetails.tsx [hunk 3]
```diff
@@ -1,18 +1,25 @@
...
 import { h, FunctionComponent, Fragment } from "preact";
 import { BookingFormPage } from "../../../Typings/BookingFormPage";
+import { useEventStore } from "../../../Hooks/useEventStore";
+import { useTimeslotStore } from "../../../Hooks/useTimeslotStore";
 import { Button } from "../../Common/Button";
 import { TextStyle } from "../../Common/TextStyle";
 import { useWizardModalAction } from "../../Common/WizardModal";
 
 export const OrderDetails: FunctionComponent = () => {
+  const event = useEventStore((state) => state.event);
```
Comment (shannon-coding) [inline src/fullPageBookingForm/Components/Views/OrderDetails/OrderDetails.tsx:11]: Currently, I'm passing the event and timeslot as a prop to the OrderDetails view, I'll be attempting to lift this into the FullPage component, do you foresee any issues with this? I'll probably find out on my own today, just off the top of your head.

File: src/fullPageBookingForm/Components/Views/TimeslotSelection/useAvailabilities.ts [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+  }>({});
+  const [availabilities, setAvailabilities] = useState<FirstAvailability>({});
+  const [timeslotsByDay, setTimeslotsByDay] = useState<
+    Record<string, Availability[]>
+  >({});
+
+  const addTimeslots = () => {
+    const timeslotsToAdd: Availability[] = [];
+
+    Object.keys(availabilities).forEach((year) => {
```
Comment (shannon-coding) [inline src/fullPageBookingForm/Components/Views/TimeslotSelection/useAvailabilities.ts:35]: Is there any way to avoid this kind of nesting?

**Comment by shannon-coding** [comment]
There also seems to be a "double scroll-bar" issue on integration, I've managed to hide the page scroll bar and only show a scroll bar for the time slot cards themselves, but I don't think that's an ideal solution. Are we supposed to have the scroll bars hidden for the time slot view?

File: src/fullPageBookingForm/Components/Views/OrderDetails/OrderDetails.tsx (hunk 3 above)
Comment (shannon-coding) [reply src/fullPageBookingForm/Components/Views/OrderDetails/OrderDetails.tsx:11]: I've updated it already, don't worry about moving it for this PR, just leave as is.

File: src/fullPageBookingForm/Components/Views/TimeslotSelection/useAvailabilities.ts (hunk 4 above)
Comment (siongesteban) [reply src/fullPageBookingForm/Components/Views/TimeslotSelection/useAvailabilities.ts:35]: Yeah this block of code doesn't look good. This is actually just my initial approach so I can make things work. The data came from the api where the response is especially designed for a calendar's use, hence the deeply-nested object.
I'll try to rework it so we avoid this deeply-nested loop.

**Comment by siongesteban** [comment]
> I'm working integration now, in general I don't see the "X" at the top left of the modal on each page, were you planning on doing that on each individual page? I could be nice to make that common and put that close functionality in the "WizardModal.tsx" file.

@shannon-coding The close button should be seen in every page. Not sure why it doesn't show in your side. If that's the case, the close button should also not show in timeslot selection page. I'll look into it though.

**Comment by siongesteban** [comment]
> There also seems to be a "double scroll-bar" issue on integration, I've managed to hide the page scroll bar and only show a scroll bar for the time slot cards themselves, but I don't think that's an ideal solution. Are we supposed to have the scroll bars hidden for the time slot view?

@shannon-coding No it shouldn't be hidden. And only the wizard modal body should be scrolled.

File: src/fullPageBookingForm/Components/Common/Calendar/Calendar.tsx (hunk 1 above)
Comment (siongesteban) [reply src/fullPageBookingForm/Components/Common/Calendar/Calendar.tsx:37]: This was intentional as I wanted to make the calendar to be both a controlled and an uncontrolled component.

Comment (shannon-coding) [reply src/fullPageBookingForm/Components/Common/Calendar/Calendar.tsx:37]: Got it, just wanted to verify.

File: src/typings/Event.ts [hunk 5]
```diff
@@ -11,6 +11,7 @@ export type EventDBO = {
   description: string;
   defaultEmailTemplateId?: IObjectID;
   endsAt?: Date;
+  featuredImageUrl: string | null;
```
Comment (shannon-coding) [inline src/typings/Event.ts:14]: So I see this has been added, however it's coming through as the image id. Are we going to update the API to ensure this gets sent as a URL?

Comment (siongesteban) [reply src/typings/Event.ts:14]: The API has been updated for this, just forgot to create a PR. Will create now.

File: src/fullPageBookingForm/Components/Views/TimeslotSelection/useAvailabilities.ts (hunk 4 above)
Comment (siongesteban) [reply src/fullPageBookingForm/Components/Views/TimeslotSelection/useAvailabilities.ts]: Updated this with recursion.

**Comment by shannon-coding** [comment]
> > I'm working integration now, in general I don't see the "X" at the top left of the modal on each page, were you planning on doing that on each individual page? I could be nice to make that common and put that close functionality in the "WizardModal.tsx" file.
> 
> @shannon-coding The close button should be seen in every page. Not sure why it doesn't show in your side. If that's the case, the close button should also not show in timeslot selection page. I'll look into it though.
> 
> EDIT:
> @shannon-coding Are you still having this problem? Btw this is how the close button should look across all the page.
> https://share.getcloudapp.com/Z4uoZO1X

So here's my latest on integration: https://share.getcloudapp.com/o0u6xD9N

I do see that you've fixed the double scroll bar issue. Some of the main issues I still see are:
1. Widget loads below shopify page, we should probably hide that on initial load until the button is clicked.

**Comment by siongesteban** [comment]
@shannon-coding This can be merged now. All the UI fixes is in #111.

**Pull request merged**